### DISCUSSION TITLE
Initial design of IWFM and IDC Migration to GitHub

### DISCUSSION THREAD

**Discussion by dezidizon** [CVWRSM/cvwrsm#110] [Ideas]
## Compilers to Consider:
- Intel Fortran Classic (ifort) - Only guaranteed working compiler, phased out once ifx is working
- Intel Fortran OneAPI / LLVM (ifx) - Intended future of project, currently has bugs. Ongoing bug reports with Intel
- ~gfortran~ - Currently not considered due to performance / compiler-specific optimizations
- ~flang~ - Currently not considered due to performance / compiler-specific optimizations

## Build Tools to Consider:
- MSBuild
- [Meson](https://mesonbuild.com/)
- [CMake](https://cmake.org/)
- [Fortran Package Manager (fpm)](https://fpm.fortran-lang.org/index.html)

### MSBuild
MSBuild is the native, XML-based build system contained within Visual Studio. It is available as a standalone CLI tool from Microsoft. Downsides to MSBuild notably include difficulty in making changes without the Visual Studio user interface, as well as poor support from non-Visual Studio IDEs and in environments without Visual Studio installed. For CI environments such as Github actions, this is a downside.

### Meson
Meson is a modern build system with support for several languages (C, C++, D, Fortran, Java, Rust), strong multi-platform support, and support across many IDEs. Notably, Meson can export MSBuild format project files for Visual Studio, as well as having native integration into Jetbrains IDEs, Eclipse, Gnome Builder, KDevelop, and vscode [via extension].[Reference](https://mesonbuild.com/IDE-integration.html). Meson is designed to be faster and easier to use than other build options such as CMake, and has reached moderately wide usage within open-source applications using C and C++. [Reference](https://mesonbuild.com/Users.html)

### CMake
CMake is a common and widely used build tool for C, C++, and Fortran projects. Due to the prevalence and age of CMake, it is widely supported, including by Visual Studio. [Reference](https://learn.microsoft.com/en-us/cpp/build/cmake-projects-in-visual-studio?view=msvc-170) Notable downsides to CMake include a complex, and difficult to understand syntax, often making it challenging to work with for developers not already experienced in CMake.

### Fortran Package Manager
The Fortran Package Manager appears to be an attempt to provide a system similar to Rust's cargo or Python's pip by the Fortran Language. Notably, it includes package management functionality and package management / artifact management support. Despite being provided by the Fortran language developers, it appears to be a relatively immature tool. The first Alpha releases of the tool date back to 2020, and full support for Package Registries was only added in mid-2023. IDE support for FPM appears to be minimal or non-existent, which makes it not a good choice at this time. As FPM is under active development, it should continue to mature, IDE support will improve, and in several years it may make sense to use FPM.

### Recommendation
Meson is the recommended choice due to ease of use, as well as wide support across IDEs and various platforms.

## Repository Organization Layout:
- Monorepo
- Subrepo
- Separate Repositories with Artifact Management

### Monorepo
A monorepo design would be the simplest. All of the IWFM / IDC components would be included as separate modules within one Git repo. This vastly simplifies repository management and coordination, as there is only one repository. Each distinct component within IWFM / IDC would remain separate, built independently, and be capable of producing separate artifacts. Release branches could be created to make specialty releases for specific uses. One important note with a monorepo is that every component would be built separately, but at the same time. Functionally, everything would be versioned together. This design has the least amount of Devops related overhead once implemented, as all the code lives in one place and is easy to obtain with a single `git clone`.

### Subrepo
There are many ways to organize a "subrepo" design, but ultimately, it would be intended for this to functionally behave similarly to the Monorepo from a development perspective. At build time, all of the subrepos would be cloned, and all code within scope would be built from source. There is additional overhead to properly keep the subrepos up-to-date within the top-level projects that include them.

#### Identical Repo Layout to SVN
All repositories in SVN would become Git repositories.

#### All Components in a Subrepo, top level applications separate
This would place all components within one large repository, and separate out the top level application interfaces (such as IWFM / IDC).

### Separate Repositories with Artifact Management
This is very similar to a standard Devops workflow in other environments and other languages, such as Java, C#, Python, Javascript, Rust, or Ruby. Each component of IWFM / IDC would be built separately, and the binary uploaded to an artifact management system. Each component that depends on a prior component at compile / link time would download the prior component from artifact management, and then use that artifact to build. This provides the most flexible design, and is considered best practices in other languages. Due to the poor support for artifact management systems within the Fortran ecosystem, this is not an ideal design for Fortran. As FPM matures, this may become a better option in the future. It may be possible to make this work, however it will likely have limited or no support in IDEs.

### Recommendation
For simplicity, ease of use, ease of development, and ease of making changes with cross-cutting concerns across several modules, the Monorepo layout is recommended. With separate modules within the Monorepo, it should be possible to maintain most benefits of the current SVN design. Additionally, it is possible to migrate to a Subrepo layout in the future by breaking up the monorepo if necessary.

**Comment by stephen-rma** [comment]
After the meeting today, we have decided to pursue:
- ifort + ifx (eventually in the future, only ifx)
- Meson
   - fpm may still be considered in the future when it matures and gains IDE support, but that is a decision to be made later
- Single Mono-subrepo with all components, application front-ends separate
- The mono-subrepo will represent history as many distinct branches representing each SVN repo

**Reply by stephen-rma** [reply]
An interesting note about OneAPI GPU support: https://www.oneapi.io/blog/bringing-nvidia-and-amd-support-to-oneapi/

There is support for AMD and NVidia GPUs via a 3rd party add-in for OneAPI. Unfortunately it appears it may be limited to C / C++ for the moment.
